k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapResponseParser.java

```diff
@@ -354,26 +354,32 @@ private Object parseLiteral() throws IOException {
         if (response.getCallback() != null) {
             FixedLengthInputStream fixed = new FixedLengthInputStream(inputStream, size);
 
+            Exception callbackException = null;
             Object result = null;
             try {
                 result = response.getCallback().foundLiteral(response, fixed);
             } catch (IOException e) {
-                // Pass IOExceptions through
                 throw e;
             } catch (Exception e) {
-                // Catch everything else and save it for later.
-                exception = e;
+                callbackException = e;
             }
 
-            // Check if only some of the literal data was read
-            int available = fixed.available();
-            if (available > 0 && available != size) {
-                // If so, skip the rest
-                while (fixed.available() > 0) {
-                    fixed.skip(fixed.available());
+            boolean someDataWasRead = fixed.available() != size;
+            if (someDataWasRead) {
+                if (result == null && callbackException == null) {
+                    throw new AssertionError("Callback consumed some data but returned no result");
                 }
+
+                fixed.skipRemaining();
             }
 
+            if (callbackException != null) {
+                if (exception == null) {
+                    exception = callbackException;
+                }
+                return "EXCEPTION";
```

Comment:
This is horrible, firstly because EXCEPTION could technically be a valid response. 

And you never test this behaviour or even use this magic value to see whether `exception` should be checked.

I'm not going to reject the changes, but at the least we should track this monstrosity and refactor it in the future